crates/cosmos/cosmos-client-components/src/impls/queries/client_state.rs

```diff
@@ -15,6 +17,39 @@ use crate::types::tendermint::TendermintClientState;
 
 pub struct QueryCosmosClientStateFromChainHandle;
 
+#[async_trait]
+impl<Chain, Counterparty> ClientStatesQuerier<Chain, Counterparty>
+    for QueryCosmosClientStateFromChainHandle
+where
+    Chain: HasIbcChainTypes<Counterparty, ClientId = ClientId> + HasBlockingChainHandle,
+    Counterparty: HasClientStateType<Chain, ClientState = TendermintClientState>,
+{
+    async fn query_client_states(
+        chain: &Chain,
+    ) -> Result<Vec<(ClientId, TendermintClientState)>, Chain::Error> {
+        chain
+            .with_blocking_chain_handle(move |chain_handle| {
+                let clients = chain_handle
+                    .query_clients(QueryClientStatesRequest {
+                        pagination: Some(PageRequest::all()),
+                    })
+                    .map_err(Chain::raise_error)?;
+
+                let clients = clients
+                    .into_iter()
+                    .map(|client| match client.client_state {
+                        AnyClientState::Tendermint(client_state) => {
```

Comment:
We would need a variant of chain components with unknown counterparty, and specify a type-erased client state type for CLI-only queries. Basically, something like:

```
pub struct AnyCounterparty;

impl<Chain> HasClientStateType<Chain> for AnyCounterparty {
    type ClientState = AnyClientState;
}
```

The `AnyCounterparty` type would be used as the counterparty only in the CLI, and chain contexts such as `CosmosChain` would need to implement query methods with `AnyCounterparty` being the counterparty.

I will create an issue for this.

Reply:
Why only in the CLI? In v1 we also scan all clients, for example to start the client refresh workers. I guess we could change the approach to use `Counterparty`, one for each other chain in config but was wondering why the CLI restriction.

Reply:
If we implement client scanning in the relay context, it would query and filter out client types that are not for the particular relay context. For example, a Cosmos-to-Cosmos relay context would only be interested in scanning for all clients of type Tendermint. In parallel, a Cosmos-to-Sovereign relay context would be scanning for all clients of type Sovereign.

It is only in some CLI commands that the counterparty chain ID is not specified, so we don't know what client type the user expects, and thus we have to support multiple client types.

That being said, I think it may be a good idea for the CLI to accept optional arguments like `--counterparty-chain-id` or `--client-type`, in the case when the user do _know_ what counterparty chain or client type that they expects. That way, if the queried client state is of different types, it would just be filtered away or shown as client type mismatch error.

Reply:
I just saw that in this CLI command, there is indeed an optional `--reference-chain` parameter. So in case that is specified, we would first determine the type of the reference chain before performing the client-specific query. Otherwise if it is not specified, we would use `AnyCounterparty` to query for all client types.

Reply:
Filed #166.

Reply:
Do we wait for #166 or can we merge this for now?

Reply:
Let's wait, I expect it'll be ready shortly.

Reply:
Actually let's merge this now and do the rest in another PR.

Reply:
Actually, let's merge this already and do `AnyCounterparty` in a follow-up PR.